polus-graph-pyramid-builder-plugin/src/main.py

```diff
@@ -111,66 +103,373 @@ def bin_data(data,column_names):
+            rows = (SortedFeats[ind2]/bincount).astype(np.uint8)   # calculate row from linear index
+            cols = np.mod(SortedFeats[ind2],bincount)              # calculate column from linear index
             counts = np.diff(ind2)                           # calculate the number of values in each bin
-            bins[feat1,feat2,rows[0],cols[0]] = ind2[0] + 1
-            bins[feat1,feat2,rows[1:],cols[1:]] = counts
-            linear_index.append([feat1,feat2])
-    return bins, bin_stats, linear_index
+            bins[i,rows[0],cols[0]] = ind2[0] + 1 
+            bins[i,rows[1:],cols[1:]] = counts 
+            graph_index.append([feat1,feat2])
+            i = i + 1
 
-""" 2. Plot Generation """
+    return bins, bin_stats, graph_index, graph_dict, yaxis
 
-# Tick formatting to mimick D3
-def format_ticks(fmin,fmax,nticks):
-    """ Generate tick labels
+def transform_data_log(data,column_names):
+    """ Bin the data
+    
+    Data from a pandas Dataframe is binned in two dimensions in a logarithmic scale. 
+    Binning is performed by binning data in one column along one axis and another column 
+    is binned along theother axis. All combinations of columns are binned without repeats 
+    or transposition. 
 
-    Polus Plots uses D3 to generate the plots. This function tries to mimic
-    the formatting of tick labels. Tick labels have a fixed width, and in
-    place of using scientific notation a scale prefix is appended to the end
-    of the number. See _prefix comments to see the suffixes that are used.
+    To figure out which bin the data belongs in is solved by using the geometric sequence
+    where:
+    a(n) = a1*(r^(n-1))
+        n = number of bins
+        a1 = first number in range (scale factor)
+        r = common ratio
+        where r (bin width) is calculated by the Freedman Diaconis Rule: 
+            https://en.wikipedia.org/wiki/Freedman–Diaconis_rule
+
+    Every column of data falls into three different categories:
+    1) dfzero2pos: All the numbers in the column are >= 0
+    2) dfneg2zero: All the numbers in the column are <= 0
+    3) dfneg2pos: The numbers in the column are either >=0 or <= 0 
+                    The number of bins, alpha, and ratio, vary to the left and right of the
+                    the y axis based on the range of the negative numbers and range of 
+                    postive numbers, respectively.
+            
+    Inputs:
+        data - A pandas Dataframe, with nfeats number of columns
+        column_names - Names of Dataframe columns
+    Outputs:
+        bins - A numpy matrix that has shape (nfeats,nfeats,bincount,bincount)
+        bin_feats - A list containing the minimum and maximum values of each column
+        linear_index - Numeric value of column index from original csv
+    """
+    #Transforms the DataFrame that Range from a Negative Number to a Positive Number
+    def dfneg2pos(data, alphavals, datmin, datmax):
```

Comment:
As discussed in Slack, I think for each of your domain functions (`dfneg2pos`, `dfneg2zero`, `dfzero2pos`) the for loop over the columns can be vectorized.

I also think that these functions can be almost completely eliminated with a few short if statements, which will make the much easier to read and modify in the future.

Please investigate.